src/DsaColor.cc

```diff
@@ -0,0 +1,197 @@
+#include "sea_dsa/DsaColor.hh"
+
+#include <fstream>
+#include <iostream> // remove after debugging
+
+using namespace sea_dsa;
+using namespace llvm;
+
+// Color class
+
+Color::Color(): m_r(0), m_g(0), m_b(0) {
+  randColor();
+}
+
+Color::Color(int r, int g, int b) : m_r(r), m_g(g), m_b(b) {};
+
+void Color::randColor() { // it would be nice to not generate the same colors all the time
+  m_r = std::rand() % 255;
+  m_g = std::rand() % 255;
+  m_b = std::rand() % 255;
+}
+
+std::string Color::stringColor() const {
+  std::ostringstream stringStream;
+
+  stringStream << "\"#" << std::hex << m_r << m_g << m_b << "\"";
+  return stringStream.str();
+}
+
+
+// What I actually want is a constructor that copies a graph but adds the two fields!
+// Actually the copy is not necessary, I just want a pointer to the graph.
+// ColorGraph class
+ColoredGraph::ColoredGraph(Graph &g, ColorMap &colorM, SafeNodeSet &safe)
+    : m_g(g), m_color(colorM), m_safe(safe) {};
+
+
+std::string ColoredGraph::getColorNode(const Node *n) const {
+
+  std::string color_string;
+  raw_string_ostream OS(color_string);
+
+  auto it = m_color.find(n);
+
+  // if it is not colored, it means that it was not mapped --> return gray
+  if(it != m_color.end()){
+    return it->getSecond().stringColor();
+  }
+  else{
+    OS << "grey";
+    return OS.str();
+  }
+}
+
+bool ColoredGraph::isSafeNode(const Node *n) const {  return m_safe.count(n) > 0; }
+
+// Coloring functions
+
+std::unique_ptr<Graph> cloneGraph(const llvm::DataLayout &dl,
+                                  Graph::SetFactory &sf,const Graph &g) {
+  std::unique_ptr<Graph> new_g( new Graph( dl, sf, g.isFlat()));
+  new_g->import(g, true /*copy all parameters*/);
+  return std::move(new_g);
+}
+
+// TODO: check Node.isModified() later to decide what to copy
+
+bool GraphExplorer::isSafeCell(SafeCellSet &f_safe, const Cell *c) {
+  return f_safe.count(c) > 0;
+}
+bool GraphExplorer::isSafeNode(SafeNodeSet &f_safe, const Node *n) {
+  return f_safe.count(n) > 0;
+}
+
+void GraphExplorer::mark_cells_graph(Graph &g, const Function &F,
```

Comment:
I'm a bit confused with **cell** being marked. A cell is a pointer to a node. In a picture, cells are represented by arrows. I thought that we would color nodes, and links in the nodes. In the picture, nodes are the boxes, and links/fields are boxes inside those boxes.

Reply:
Yes, you are right, I will change the implementation to make it in terms of nodes.

Reply:
After reviewing what I wanted to do, I think it is more useful to have this function in terms of cells of the graph, since the mapping between caller and callee is in terms of cells. However I think that it is more accurate if renaming this function by `explore_cells_graph`.

The public method of `GraphExplorer`, `colorGraph`, outputs the results in terms of nodes of the graph.

Reply:
But you cannot color a cell. Cells cannot have any properties attached to them. They are just pointers to Nodes. So something is off somewhere

Reply:
I changed the way in which the graphs are explored, now everything is in terms of nodes.